quill-core/src/main/scala/io/getquill/norm/AdHocReduction.scala

```diff
@@ -52,6 +44,26 @@ object AdHocReduction {
       case FlatMap(a, b, UnionAll(c, d)) =>
         Some(UnionAll(FlatMap(a, b, c), FlatMap(a, b, d)))
 
+      // query(...).take(n).map(d => e) =>
+      //   query(...).map(d => e).take(n)
+      case Map(Take(fm: FlatMap, n), ma, mb) =>
+        Some(Take(Map(fm, ma, mb), n))
+
+      // query(...).drop(n).map(d => e) =>
+      //   query(...).map(d => e).drop(n)
+      case Map(Drop(fm: FlatMap, n), ma, mb) =>
+        Some(Drop(Map(fm, ma, mb), n))
+
+      // query(...).take(n).drop(m).map(d => e) =>
```

Comment:
Previous just for single take or single drop, these two for combined. Yes this seems somehow like workaround(most of query compilation bugs now are somehow connected to nested query generation imho). 
For comp translates into flatMap().?.map(), where ? is combinations of take/drop. We don't want them to compile into nested queries. That's why I reorder theses and later in SqlQuery query calls `flatten` on nested query but no in the whole. wdyt?